src/ast/visitor.rs

```diff
@@ -166,6 +210,84 @@ pub trait Visitor {
     }
 }
 
+/// A visitor that can be used to mutate an AST tree.
```

Comment:
One thing that might be worth considering is to change this from mutating to rewriting by taking ownership.

Having such a signature  could be more efficient in cases where an owned value is needed, for example, rather than

```rust
impl VisitorMut for Replacer {
  type Break = ();
  fn post_visit_expr(&mut self, expr: &mut Expr) -> ControlFlow<Self::Break> {
    if let Expr::Identifier(Ident{ value, ..}) = expr {
      // rewrite the expr to function call
      *expr = Expr::FunctionCall(name: value.clone(), ...)
    }
    ControlFlow::Continue(())
  }
}
```

One could write something like the following (reusing the allocated value):

```rust
impl VisitorRewrite for Rewriter {
  fn post_visit_expr(&mut self, mut expr: Expr) -> Result<Expr, Break> {
    if let Expr::Identifier(Ident{ value, ..}) = expr {
      Ok(Expr::FunctionCall(name: value, ...)
    } else {
      Ok(expr)
    }
}
```

Reply:
I think copying all the potentially large elements in the AST multiple times in an out of functions would actually be less performant. And that would require to put something inside the AST while the function has the ownership of the elements, requiring everything to be `Default`, taking time to instantiate values that will never be read, and leaving the statement in an unexpected state if the function panics.

Reply:
In the example you give, you don't need to clone the ident value, you can just [`take`](https://doc.rust-lang.org/std/mem/fn.take.html) it.

Reply:
FWIW the sizes are

```
#[test]
fn test() {
    assert_eq!(std::mem::size_of::<Statement>(), 712);
    assert_eq!(std::mem::size_of::<Query>(), 536);
    assert_eq!(std::mem::size_of::<Expr>(), 144);
}
```

These are likely large enough that moving them in and out of functions may have a non-trivial overhead.

I also confirmed that Rust isn't able to elide these copies in the general case - https://rust.godbolt.org/z/3W8YfWM6q

That being said Rust inlines very aggressively and inlining the visitor body may eliminate this.

I don't feel strongly about either solution, particularly in the absence of any benchmarks either way. In either case a rewrite will always result in a memcpy, passing by reference avoids needing the compiler to inline and elide the move in the non-mutating case, but passing by value may help it better optimise other cases, I don't really know :sweat_smile:

Reply:
I made a small toy version to compare the two, the by-value version is 4x slower even with a small struct: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2abc5db67462f0516fd37e5d865a9563